sycl/doc/extensions/proposed/sycl_ext_oneapi_free_function_kernels.asciidoc

```diff
@@ -0,0 +1,674 @@
+`bundle_state::executable`.
+
+_Preconditions_: The address `Func` must be the address of some free function
+kernel that is defined in the calling application.
+
+_Returns:_ If the kernel whose address is `Func` resides in this kernel bundle,
+returns the `kernel` object representing that kernel.
+
+_Throws_: An `exception` with the error code `errc::invalid` if the kernel with
+address `Func` does not reside in this kernel bundle.
+|====
+
+=== Behavior with kernel bundle functions in the core SYCL specification
+
+Free function kernels that are defined by the application have a corresponding
+kernel identifier (`kernel_id`) and are contained by the device images in the
+SYCL application.
+This section defines the ramifications this has on the kernel bundle functions
+defined by the core SYCL specification.
+
+* The function `get_kernel_ids()` returns the kernel identifiers for any free
+  function kernels defined by the application, in addition to identifiers for
+  any kernels defined as lambda expressions or named kernel objects.
+
+* The kernel bundle returned by
+  `get_kernel_bundle(const context&, const std::vector<device>& devs)` contains
+  all of the free function kernels defined by the application that are
+  compatible with at least one of the devices in `devs`, in addition to all of
+  the kernels defined as lambda expressions or named kernel objects that are
+  compatible with one of these devices.
+
+* The function `has_kernel_bundle(const context&, const std::vector<device>&)`
+  considers free function kernels defined by the application when computing its
+  return value.
+
+The information descriptor `info::kernel::num_args` may be used to query a
+`kernel` object that represents a free function kernel.
+The return value tells the number of formal parameters in the function's
+definition.
+
+=== Enqueuing a free function kernel and setting parameter values
+
+Once the application obtains a `kernel` object for a free function kernel, it
+can enqueue the kernel to a device using any of the SYCL functions that allow
+a kernel to be enqueued via a `kernel` object.
+The application must enqueue the free function kernel according to its type.
+For example, a free function kernel defined via `SYCL_EXT_ONEAPI_RANGE_KERNEL`
+can be enqueued by calling the `handler::parallel_for` overload taking a
+`range`.
+A free function kernel defined via `SYCL_EXT_ONEAPI_ND_RANGE_KERNEL` can be
+enqueued by calling the `handler::parallel_for` overload taking an `nd_range`.
+A free function kernel defined via `SYCL_EXT_ONEAPI_SINGLE_TASK_KERNEL` can be
+enqueued by calling `handler::single_task`.
+
+Attempting to enqueue a free function kernel using a mechanism that does not
+match its type results in undefined behavior.
+Attempting to enqueue a free function kernel with a `range` or `nd_range` whose
+dimensionality does not match the free function kernel definition results in
+undefined behavior.
```

Comment:
I was thinking that by having three different macros we can provide better error messages.  For example, we can emit a compile-time error if a kernel decorated with `SYCL_EXT_ONEAPI_RANGE_KERNEL` calls the function `sycl::ext::oneapi::this_kernel::get_nd_item`.  We can also throw an exception if a kernel decorated with `SYCL_EXT_ONEAPI_RANGE_KERNEL` is enqueued as an nd-range kernel.

Note that we can diagnose these errors even if the spec says the behavior is UB.  We always have the option to diagnose UB with a helpful error.

Reply:
> For example, we can emit a compile-time error if a kernel decorated with `SYCL_EXT_ONEAPI_RANGE_KERNEL` calls the function `sycl::ext::oneapi::this_kernel::get_nd_item`.

Is this a situation where we need to emit an error? I would expect any SYCL kernel to work for any combination of kernel invocation API and kernel identifiers free functions. nd_item is a wrapper around kernel identifiers, which are available to any kernel regardless of parallel_for overload. Current SYCL API limitations on lambda call operator parameter type are artificial. They are useful for code optimization: compiler has to allocate and initialize id/item/nd_item lambda argument. In this case, it's created by user via free function, so we don't need to optimize it.

Reply:
I do think an error would be helpful.  The specification for the free function queries says their behavior is undefined if they are called from a kernel that is launched in an incompatible way:

> It is the user’s responsibility to ensure that these functions are called in a manner that is compatible with the kernel’s launch parameters, as detailed in the definition of each function. Calling these functions from an incompatible kernel results in undefined behavior.

(https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_free_function_queries.asciidoc#accessing-instances-of-special-sycl-classes)

If we don't say it is UB, then we would need to clarify the spec to explain what is returned in the mismatched cases.  For example, what is the work-group size for a simple "range" parallel-for?  This seems worse than saying the behavior is UB and diagnosing an error when possible.

Reply:
> If we don't say it is UB, then we would need to clarify the spec to explain what is returned in the mismatched cases.

There are no mismatches cases because there is no SYCL kernel argument we have to "match" (or to be compatible with). This argument exists only for call operator of functional object, but this extension defines alternative API, which we don't need to "match".

> For example, what is the work-group size for a simple "range" parallel-for?

I expect it to be defined by the implementation. AFAIK, all supported backends provide access to work-group size on device side.
On the host side, when not set explicitly OpenCL back-end chooses the work-group size automatically. Other backends has API to query recommended work-group size, which can be set explicitly.

Reply:
> There are no mismatches cases because there is no SYCL kernel argument we have to "match"

By "mismatched" I meant a mismatch between the way the kernel is launched and the queries the kernel makes.  For example, launching the kernel as a simple "range" kernel and calling `this_kernel::get_nd_item`.

> I expect it to be defined by the implementation. AFAIK, all supported backends provide access to work-group size on device side.

I really feel this would open a can of worms that would make our life difficult in the future.  For example, what if the application launches the kernel as a simple "range" and then uses `this_kernel::get_nd_item` to get a `group` object that it uses to call one of the group functions defined in [section 4.17.2](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:group-functions) or [4.17.3](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:algorithms)?  I really don't think we want to say this is supported.  If we were required to support this, I think it would limit our implementation choices in the future.

I also think this presents a confusing story to our users.  @Pennycook has been clarifying the SYCL spec to say that simple "range" kernels have no concept of work-group.  Therefore, it does not make sense to allow "range" kernels to call `this_kernel::get_nd_item`.